### PR TITLE
[feat]: [CDS-47025] Asg Rolling Deploy - Manager Changes

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
Added Manager Side Implementation for ASG NG Swimlane Rolling Deploy step. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)

### CODE REVIEW

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    return asgStepCommonHelper.startChainLink(this, ambiance, stepParameters);
+  }
+
+  @Override
+  public TaskChainResponse executeNextLinkWithSecurityContext(Ambiance ambiance, StepElementParameters stepParameters,
+      StepInputPackage inputPackage, PassThroughData passThroughData, ThrowingSupplier<ResponseData> responseSupplier)
+      throws Exception {
+    log.info("Calling executeNextLink");
+    return asgStepCommonHelper.executeNextLinkRolling(
+            this, ambiance, stepParameters, passThroughData, responseSupplier, asgStepHelper);
```
Comment (vitsafronovici) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java:72]: no need to pass `asgStepHelper ` it's already defined in AsgStepCommonHelper

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepCommonHelper.java [hunk 2]
```diff
@@ -344,4 +360,76 @@ public AsgInfraConfig getAsgInfraConfig(InfrastructureOutcome infrastructure, Am
     NGAccess ngAccess = AmbianceUtils.getNgAccess(ambiance);
     return asgEntityHelper.getAsgInfraConfig(infrastructure, ngAccess);
   }
+
+  public TaskChainResponse executeNextLinkRolling(AsgStepExecutor asgStepExecutor, Ambiance ambiance,
```
Comment (vitsafronovici) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepCommonHelper.java:364]: why does the method name have `NextLink`?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java (hunk 1 above)
Comment (lovish1999) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java:72]: Okay, editing that.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepCommonHelper.java (hunk 2 above)
Comment (lovish1999) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepCommonHelper.java:364]: As it is the next link task for rolling deploy step. Just followed the naming used in ECS NG.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java` line 104 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJnAM954aLvQlcktVMp:-NJnAM954aLvQlcktVMq:b-4pfyih) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java#L104)):*
> ```Java
> 
>   @Override
>   public TaskChainResponse executeAsgPrepareRollbackTask(Ambiance ambiance, StepElementParameters stepElementParameters,
> ```

the method name is confusing at it may refer to Rollback. We can keep this in sync with PrepareRollbackDataTask

**Review by vitsafronovici — APPROVED** [review]
> Reviewed 18 of 29 files at r1, 3 of 6 files at r2, 8 of 20 files at r3, 9 of 14 files at r4, 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 44 file(s) to review. 1 of 2 approvals obtained.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingRollbackDataOutcome.java` line 32 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJnAtfW1VYLsQqOf5SA:-NJnAtfW1VYLsQqOf5SB:bcomi1l) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingRollbackDataOutcome.java#L32)):*
> ```Java
>   String asgName;
>   Map<String, List<String>> asgStoreManifestsContent;
>   private boolean isFirstDeployment;
> ```

what does this field signify? If there is no ASG, we'll receive no manifest content, right?

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java` line 51 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoRdVy-1vvUJyt8Gp6:-NJoRdVz4VyhaPBajaPK:bxf55qk) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java#L51)):*
> ```Java
> 
>   private static final String ASG_ROLLING_DEPLOY_COMMAND_NAME = "AsgRollingDeploy";
>   private static final String ASG_PREPARE_ROLLBACK_COMMAND_NAME = "AsgPrepareRollback";
> ```

Let's rename everywhere prepare_rollback to prepare_rollback_data (variables, values, classes, etc). Else this would be confusing for maintainer.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepHelper.java` line 42 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXAt59AyCkLsU92dG:-NJoXAt59AyCkLsU92dH:b-m8kgro) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepHelper.java#L42)):*
> ```Java
> 
>     // Check if ASG Manifests are empty
>     if (isEmpty(asgManifests)) {
> ```

how would we make sure if ASG Configuration and Launch Template, both are available?

**Comment by arvind-choudhary-h** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java` line 65 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXaOBDXrN521-P0Lj:-NJoXaOC7q3mQHpoKLWm:b-l6mkjh) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java#L65)):*
> ```Java
>   public final String ASG_CANARY_DELETE_OUTCOME = "asgCanaryDeleteOutcome";
>   public final String ASG_CANARY_DEPLOY_OUTCOME = "asgCanaryDeployOutcome";
>   public final String ASG_ROLLING_ROLLBACK_OUTCOME = "asgRollingRollbackOutcome";
> ```

rename

**Comment by arvind-choudhary-h** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgPrepareRollbackDataResult.java` line 24 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXnkN9opxlmPECBE8:-NJoXnkN9opxlmPECBE9:b-mchm1z) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgPrepareRollbackDataResult.java#L24)):*
> ```Java
>   String asgName;
>   Map<String, List<String>> asgStoreManifestsContent;
>   private boolean isFirstDeployment;
> ```

do we need this?

**Comment by arvind-choudhary-h** [comment]
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgRollingDeployResponse.java` line 28 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXw_03QBbx4V3i9ts:-NJoXw_03QBbx4V3i9tt:b3vwz1u) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgRollingDeployResponse.java#L28)):*
> ```Java
>   CommandExecutionStatus commandExecutionStatus;
>   String errorMessage;
>   // AsgRollingDeployResult asgRollingDeployResult;
> ```

we'll need this.

**Review by arvind-choudhary-h — COMMENTED** [review]
> Reviewed 18 of 29 files at r1, 3 of 6 files at r2, 9 of 20 files at r3, 9 of 14 files at r4, 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 1 of 2 approvals obtained.

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 6. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java` line 51 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoRdVy-1vvUJyt8Gp6:-NJsq-YQD9_U7hs5IVnf:b32w1t) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java#L51)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

Let's rename everywhere prepare_rollback to prepare_rollback_data (variables, values, classes, etc). Else this would be confusing for maintainer.
</blockquote></details>

Ok
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java` line 104 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJnAM954aLvQlcktVMp:-NJsq7s32vpW9oeAWO7d:b32wtd) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java#L104)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

the method name is confusing at it may refer to Rollback. We can keep this in sync with PrepareRollbackDataTask
</blockquote></details>

ok
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepHelper.java` line 42 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXAt59AyCkLsU92dG:-NJsqAM69svMK4X7OtLL:bt0dp71) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgStepHelper.java#L42)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

how would we make sure if ASG Configuration and Launch Template, both are available?
</blockquote></details>

This code handles that. its just after the above code. 



_Code snippet:_
```Java
   // Get AsgLaunchTemplate manifests and validate
    List<ManifestOutcome> asgLaunchTemplates =
        filterManifestOutcomesByType(asgManifests, ManifestType.AsgLaunchTemplate);
    assertManifestOutcomeSizeOne(asgLaunchTemplates, ManifestType.AsgLaunchTemplate);

    // Get AsgConfiguration manifests and validate
    List<ManifestOutcome> asgConfigurations = filterManifestOutcomesByType(asgManifests, ManifestType.AsgConfiguration);
    assertManifestOutcomeSizeOne(asgConfigurations, ManifestType.AsgConfiguration);
```
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java` line 65 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXaOBDXrN521-P0Lj:-NJsqJvW-4e9u1tvoO3X:b32wtd) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java#L65)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

rename
</blockquote></details>

ok
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgPrepareRollbackDataResult.java` line 24 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXnkN9opxlmPECBE8:-NJssLXW2hO-hYfiky67:b-w5kr9t) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgPrepareRollbackDataResult.java#L24)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

do we need this?
</blockquote></details>

It is required in rolling rollback step on delegate side while deciding whether can rollback or not.    For reference: EcsRollingRollbackCommandTaskHandler.java
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgRollingDeployResponse.java` line 28 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJoXw_03QBbx4V3i9ts:-NJstQBU6NYk9WzP4qfA:b7fjxlw) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/asg/AsgRollingDeployResponse.java#L28)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

we'll need this.
</blockquote></details>

ok adding AutoScalingGroupContainer as done for asgcanarydeployresult

**Review by lovish1999 — COMMENTED** [review]
> It is required in rolling rollback step on delegate side while deciding whether can rollback or not. For reference: EcsRollingRollbackCommandTaskHandler.java

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733#-:-NJstsHb5wrTiQpPYXCh:bjoeip7)* status: 7. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingRollbackDataOutcome.java` line 32 at r5](https://reviewable.io/reviews/harness/harness-core/41733#-NJnAtfW1VYLsQqOf5SA:-NJstt_a6U7IZjfrfEp_:bm0cd0h) ([raw file](https://github.com/harness/harness-core/blob/88c2bf628ab80a6b0905cbbc6576773b56cd42cd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingRollbackDataOutcome.java#L32)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

what does this field signify? If there is no ASG, we'll receive no manifest content, right?
</blockquote></details>

It is required in rolling rollback step on delegate side while deciding whether can rollback or not. For reference: EcsRollingRollbackCommandTaskHandler.java

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java` line 51 at r6](https://reviewable.io/reviews/harness/harness-core/41733#-NJxXv_-BoJQ1F8ByHff:-NJxXv_-BoJQ1F8ByHfg:b-131v6r) ([raw file](https://github.com/harness/harness-core/blob/a303dd891122d42047e8733bb5ec4806a66d3abd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java#L51)):*
> ```Java
> 
>   private static final String ASG_ROLLING_DEPLOY_COMMAND_NAME = "AsgRollingDeploy";
>   private static final String ASG_ROLLING_PREPARE_ROLLBACK_DATA_COMMAND_NAME = "AsgRollingPrepareRollbackData";
> ```

will this same rollback data be useful for BG deployment? If yes, we can rename this to AsgPrepareRollbackData; ?



_Code quote:_
```Java
ASG_ROLLING_PREPARE_ROLLBACK_DATA_COMMAND_NAME
```

**Comment by arvind-choudhary-h** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java` line 66 at r6](https://reviewable.io/reviews/harness/harness-core/41733#-NJxYSCB2EPwi_IzcIJp:-NJxYSCB2EPwi_IzcIJq:b-p913v6) ([raw file](https://github.com/harness/harness-core/blob/a303dd891122d42047e8733bb5ec4806a66d3abd/127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java#L66)):*
> ```Java
>   public final String ASG_CANARY_DELETE_OUTCOME = "asgCanaryDeleteOutcome";
>   public final String ASG_CANARY_DEPLOY_OUTCOME = "asgCanaryDeployOutcome";
>   public final String ASG_ROLLING_PREPARE_ROLLBACK_DATA_OUTCOME = "asgRollingPrepareRollbackDataOutcome";
> ```

Can we share the rollback data between Rolling and BG deployment.?

**Review by arvind-choudhary-h — COMMENTED** [review]
> Reviewed 12 of 12 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 12 file(s) to review. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 18 of 29 files at r1, 2 of 6 files at r2, 6 of 20 files at r3, 4 of 14 files at r4, 5 of 6 files at r5, 12 of 12 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 1 of 2 approvals obtained.

**Review by lovish1999 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 8 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java` line 51 at r6](https://reviewable.io/reviews/harness/harness-core/41733#-NJxXv_-BoJQ1F8ByHff:-NJzIkzU3Mv6vO0gIyPL:b-xqhq11) ([raw file](https://github.com/harness/harness-core/blob/a303dd891122d42047e8733bb5ec4806a66d3abd/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgRollingDeployStep.java#L51)):*
> I Will have to see this when making BG changes.
I Will have to see this when making BG changes.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java` line 66 at r6](https://reviewable.io/reviews/harness/harness-core/41733#-NJxYSCB2EPwi_IzcIJp:-NJzIf6o2YZ88JdEpmv0:b-yui7z3) ([raw file](https://github.com/harness/harness-core/blob/a303dd891122d42047e8733bb5ec4806a66d3abd/127-cd-nextgen-entities/src/main/java/io/harness/cdng/stepsdependency/constants/OutcomeExpressionConstants.java#L66)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

Can we share the rollback data between Rolling and BG deployment.?
</blockquote></details>

I Will have to see this when making BG changes.

**Review by vitsafronovici — APPROVED** [review]
> Reviewed 10 of 12 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: 8 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by arvind-choudhary-h — APPROVED** [review]
> Reviewed 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41733)* status: :shipit: complete! ✓

**Pull request merged**